### ISSUE TITLE
errors in console with empty plugin example

### ISSUE THREAD

**Issue by 0xpr03** [ReSpeak/rust-ts3plugin#4]
Hi, thanks for the lib first of all.

I currently get the this errors all over the place when using the clean plugin example:
```log
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
10/22/18 22:38:54	rust-ts3plugin	Error	Error NotConnected (Can't get channel information) in in ts3interface.rs:L438	
```
![image](https://user-images.githubusercontent.com/1280209/47323614-d09bb200-d65c-11e8-9237-13140ed95c3f.png)

Any ideas what I'm doing wrong ?

**Comment by 0xpr03** [comment]
Code:

```rust
#[macro_use]
extern crate ts3plugin;
#[macro_use]
extern crate lazy_static;

use ts3plugin::*;

struct MyTsPlugin;

const PLUGIN_NAME_I: &'static str = env!("CARGO_PKG_NAME");

impl Plugin for MyTsPlugin {
    fn new(api: &mut TsApi) -> Result<Box<MyTsPlugin>, InitError> {
        api.log_or_print("Inited", PLUGIN_NAME_I, LogLevel::Info);
        Ok(Box::new(MyTsPlugin))
        // Or return Err(InitError::Failure) on failure
    }

    // Implement callbacks here

    fn shutdown(&mut self, api: &mut TsApi) {
        api.log_or_print("Shutdown", PLUGIN_NAME_I, LogLevel::Info);
    }
}

create_plugin!(
    PLUGIN_NAME_I, env!("CARGO_PKG_VERSION"), env!("CARGO_PKG_AUTHORS"), "yamba ts3 controller",
    ConfigureOffer::No, true, MyTsPlugin);
```

**Comment by Flakebi** [comment]
Hm, I’m currently using the `v3.0` branch myself and beginning with a TeamSpeak version some months ago, I don’t get the channel names anymore (kind of unfortunate for text-to-speech) and I also get some warnings.
Your problem could be related. Unfortunately I don’t have the time to look into this problem at the moment.
The rest of the functionality should work though.

If you don’t need the channel data and the log messages bother you, you can just comment them out in the source code for now.

**Edit**:
Interestingly, this happens only for some servers, but not for others.

**Comment by 0xpr03** [comment]
I'm wondering first of all as I'm not using anything, so I wouldn't expect any warnings for functions I'm not using, especially not under a library name tag.
It happens less when I'm connected, nearly gone.
Edit: Server relation could be tied to server version ?

**Comment by Flakebi** [comment]
The way in which the TeamSpeak plugin API lets you query client and channel properties is a bit suboptimal. For example you cannot get the name of a client anymore when you get the disconnected event for this client (because TeamSpeak tells you that he is not connected anymore).

But this is needed to develop e.g,. a text-to-speech plugin. So I decided to store the structures of channels and clients on a server inside this library and query the plugin API only when something changes. Plugins build on top of this can then query any kind of stored information at any time they want.

This means, even if your plugin does not use channel and client information, this library will still try to store them (for the case that you want to have them).

For the server versions, at least yesterday I did not get any warnings for one 3.4 server, but I did get warnings for another 3.4 server (though I did not watch that closely).

**Comment by 0xpr03** [comment]
Ah, so it's same problem as with server query connections. ( I've written [ts3-manager](https://github.com/0xpr03/ts3-manager) for reference ), where you only get the cID but not the real user ID on DC, so you've to store them on connect.
Well, I can understand the idea, but would personally prefer to leave this to the user. This way there's no black magic involved which leads to a situation like this.

**Comment by Flakebi** [comment]
I hope this is fixed with v0.3 (though I don’t remember all the details).
